src/e2e_test/perf_tool/pkg/run/run.go

```diff
@@ -104,139 +107,216 @@ func (r *Runner) RunExperiment(ctx context.Context, expID uuid.UUID, spec *exper
-	for i, w := range r.workloads {
-		name := spec.WorkloadSpecs[i].Name
+	eg := errgroup.Group{}
+	for _, w := range r.workloadsBySelector[selector] {
 		workload := w
 		eg.Go(func() error {
-			log.WithField("workload", name).Trace("Waiting for workload healthcheck")
+			log.WithField("workload", workload.Name()).Trace("Waiting for workload healthcheck")
 			if err := workload.WaitForHealthCheck(ctx, r.clusterCtx, spec.ClusterSpec); err != nil {
 				return err
 			}
-			log.WithField("workload", name).Trace("HealthCheck passed")
+			log.WithField("workload", workload.Name()).Trace("HealthCheck passed")
 			return nil
 		})
 	}
 
 	if err := eg.Wait(); err != nil {
-		return err
+		_ = r.stopWorkloads(selector)
+		return noCleanup, err
 	}
+	cleanup := func() { _ = r.stopWorkloads(selector) }
+	return cleanup, nil
+}
 
-	// Add a row to the results with the time when workloads were deployed.
-	row := &metrics.ResultRow{
-		Timestamp: time.Now(),
-		Name:      "workloads_deployed",
-		Value:     0.0,
-	}
-	select {
-	case <-ctx.Done():
-	case metricsResultCh <- row:
-	}
+func (r *Runner) stopVizier() error {
+	log.Info("Stopping Vizier")
+	return r.vizier.Close()
+}
 
-	// Wait for the experiment duration specified in the RunSpec.
-	// During this time, Vizier and workloads are deployed and metrics are recording.
-	dur, err := types.DurationFromProto(spec.RunSpec.Duration)
-	if err != nil {
-		return err
-	}
-	log.Infof("Experiment running for %v", dur)
-	if sleep(ctx, dur) {
+func (r *Runner) stopMetricRecorders(selector string) error {
+	log.WithField("selector", selector).Info("Stopping metric recorders")
+	mrs, ok := r.metricsBySelector[selector]
+	if !ok {
 		return nil
 	}
-	log.Info("Experiment finished tearing down")
-
-	// Teardown the experiment. First teardown metrics, then workloads, then the vizier.
-	// This is a different order than the defers would occur.
-	// Since all of the Close() methods should be idempotent, it should be fine that we call Close() twice.
-	// We report the first error from any of the Close() methods, the defers will take care of tearing down the rest.
-	log.Trace("Tearing down metric recorders")
-	for _, recorder := range mrs {
-		if err := recorder.Close(); err != nil {
+	defer func() {
+		for _, mr := range mrs {
+			mr.Close()
+		}
+	}()
+	for _, mr := range mrs {
+		// We return the first error, and the defer will cleanup the rest.
+		if err := mr.Close(); err != nil {
 			return err
 		}
 	}
+	return nil
+}
 
-	log.Trace("Tearing down workloads")
-	for _, w := range r.workloads {
+func (r *Runner) stopWorkloads(selector string) error {
+	log.WithField("selector", selector).Info("Stopping workloads")
+	ws, ok := r.workloadsBySelector[selector]
+	if !ok {
+		return nil
+	}
+	defer func() {
+		for _, w := range ws {
+			w.Close()
+		}
+	}()
+	for _, w := range ws {
+		// We return the first error, and the defer will cleanup the rest.
 		if err := w.Close(); err != nil {
 			return err
 		}
 	}
+	return nil
```

Comment:
cleaner to use `errors.Join` to just grab all errors. you can stick them in a slice and then call join.
https://pkg.go.dev/errors#Join